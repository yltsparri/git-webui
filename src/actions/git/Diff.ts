export interface Diff {
  headerLines: string[];
  fileDiffs: FileDiff[];
}

export interface FileDiff {
  header: string;
  fileModeLine: string;
  indexLine: string;
  initialFile: string;
  resultingFile: string;
  hunks: Hunk[];
}

export enum HunkPartType {
  Keep = 0,
  Remove = 1,
  Add = 2
}

export interface HunkPart {
  content: string[];
  type: HunkPartType | null;
}

export interface Hunk {
  header: string;
  parts: HunkPart[];
}

export default Diff;
