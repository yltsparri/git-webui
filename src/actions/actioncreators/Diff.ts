/*
 * Copyright 2017 Ãœlo Parri
 * Copyright 2015 Eric ALBER
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Action } from "redux";
import { ThunkDispatch } from "redux-thunk";
import Actions from "../Actions";
import { AppState, CommitViewMode } from "../AppState";
import { FileInfo } from "../git/FileInfo";
import Git from "../git/Git";
import { addMessage } from "./Messages";
import { selectNode } from "./Tree";

export const loadDiff = (commit: string) => {
  return (
    dispatch: ThunkDispatch<{}, {}, Action>,
    getState: () => AppState
  ) => {
    const state = getState() as AppState;
    const { ignoreWhitespace, options, context, fullFile } = state.diffOptions;
    let diffContext = context;
    if (fullFile) {
      diffContext = 99999999;
    }
    return Git.getDiff(commit, diffContext, ignoreWhitespace, options).then(
      response => {
        dispatch({
          type: Actions.UPDATE_COMMIT_DIFF_DATA,
          data: response.data
        });
        if (response.message) {
          dispatch(addMessage(response.message));
        }
      }
    );
  };
};

export function selectDiffViewMode(mode: CommitViewMode) {
  return (
    dispatch: ThunkDispatch<{}, {}, Action>,
    getState: () => AppState
  ) => {
    dispatch({ type: Actions.SELECT_COMMIT_VIEW_MODE, viewMode: mode });
    const state = getState() as AppState;
    const opts = state.commitTree;
    if (
      mode === CommitViewMode.Tree &&
      opts.path.length < 2 &&
      !opts.files.length
    ) {
      let root: FileInfo;
      if (opts.path.length) {
        root = opts.path[opts.path.length - 1];
      } else {
        root = {
          name: state.appData.dirName,
          size: NaN,
          objectId: state.commits.selectedCommit,
          isSymbolicLink: false,
          mode: 0,
          type: "tree",
          parent: null
        };
      }
      dispatch(selectNode(root));
    }
  };
}

export function setDiffContext(context: number) {
  return (
    dispatch: ThunkDispatch<{}, {}, Action>,
    getState: () => AppState
  ) => {
    dispatch({
      type: Actions.UPDATE_COMMIT_VIEW_DATA,
      data: { context }
    });
    const state = getState();
    if (!state.diffOptions.fullFile) {
      const commitHash = state.commits.selectedCommit;
      dispatch(loadDiff(commitHash));
    }
  };
}

export function toggleIgnoreWhiteSpace() {
  return (
    dispatch: ThunkDispatch<{}, {}, Action>,
    getState: () => AppState
  ) => {
    dispatch({ type: Actions.TOGGLE_IGNORE_WHITESPACE });
    const commitHash = getState().commits.selectedCommit;
    dispatch(loadDiff(commitHash));
  };
}

export function toggleShowFullFile() {
  return (
    dispatch: ThunkDispatch<{}, {}, Action>,
    getState: () => AppState
  ) => {
    dispatch({ type: Actions.TOGGLE_SHOW_FULL_FILE });
    const commitHash = getState().commits.selectedCommit;
    dispatch(loadDiff(commitHash));
  };
}

export function selectFile(index: number) {
  return { type: "SELECT_COMMIT_DIFF_FILE", selectedFile: index };
}
