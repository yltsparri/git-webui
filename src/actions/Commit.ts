import { FileDiff } from './git/Diff';
import FileInfo from './git/FileInfo';
import CommitInfo from './git/CommitInfo';

export enum CommitViewMode {
  Diff = 0,
  SidebySideDiff = 1,
  Tree = 2
}

export interface DiffOptions {
  ignoreWhitespace: boolean;
  context: number;
  fullFile: boolean;
  options?: Array<string>;
}

export interface CommitTree {
  path: Array<FileInfo>;
  files: Array<FileInfo>;
}

export interface CommitDiff {
  fileDiffs: Array<FileDiff>;
  selectedFile: number;
  headerLines: Array<string>;
  useSplitDiff: boolean;
  removedLinesDiff: FileDiff;
  addedLinesDiff: FileDiff;
}

export interface Path {
  key: string;
  commands: Array<{ type: string, x: number, y: number }>;
}

export interface Graph {
   paths: Array<Path>;
   circles: Array<Circle>;
}

export interface Circle {
  cx: number;
  cy: number;
  r: number;
  key: string;
}

export interface Commits {
  commits: Array<CommitInfo>;
  selectedCommit: string;
  viewMode: CommitViewMode;
  graph: Graph;
}

export default Commits;
