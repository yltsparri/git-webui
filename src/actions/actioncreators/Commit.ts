/*
 * Copyright 2017 Ãœlo Parri
 * Copyright 2015 Eric ALBER
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Action } from "redux";
import { ThunkDispatch } from "redux-thunk";
import Actions from "../Actions";
import { AppState, CommitViewMode } from "../AppState";
import { CommitInfo } from "../git/CommitInfo";
import { FileInfo } from "../git/FileInfo";
import { loadDiff } from "./Diff";
import { selectNode } from "./Tree";

export function commitSelected(commit: CommitInfo) {
  return (
    dispatch: ThunkDispatch<{}, {}, Action>,
    getState: () => AppState
  ) => {
    let state = getState();
    const root: FileInfo = {
      name: state.appData.dirName,
      size: NaN,
      objectId: commit.hash,
      isSymbolicLink: false,
      mode: 0,
      type: "tree",
      parent: undefined
    };
    dispatch({ type: Actions.SELECT_COMMIT, selectedCommit: commit.hash });

    dispatch({
      type: Actions.UPDATE_COMMIT_VIEW_DATA,
      data: { commitHash: commit.hash, path: [root], files: [] }
    });
    state = getState();
    if (state.commits.viewMode === CommitViewMode.Diff) {
      dispatch(loadDiff(commit.hash));
    } else if (state.commits.viewMode === CommitViewMode.Tree) {
      dispatch(selectNode(root));
    }
  };
}
