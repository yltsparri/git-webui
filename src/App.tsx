import * as React from "react";
import { connect } from "react-redux";
import { AppMode, AppState } from "./actions/AppState";
import RemoteView from "./components/RemoteView";
import Explore from "./containers/Explore";
import HistoryView from "./containers/History";
import ModalMessage from "./containers/ModalMessage";
import Sidebar from "./containers/Sidebar";
import "./css/bootstrap.css";
import "./css/main.css";

interface IndexProps {
  mode: AppMode;
  repo: string;
}

export class App extends React.Component<IndexProps> {
  public render() {
    return (
      <div className="root">
        <Sidebar />
        <div id="main-view">
          <ModalMessage />
          {this.getMainViewContents()}
        </div>
      </div>
    );
  }

  private getMainViewContents = () => {
    switch (this.props.mode) {
      case AppMode.RemoteAccess:
        return <RemoteView repo={this.props.repo} />;
      case AppMode.Workspace:
        return <div />;
      case AppMode.Explore:
        return <Explore />;
      default:
        return <HistoryView />;
    }
  }
}

const ConnectedApp = connect<IndexProps, {}, {}, AppState>(
  (state: AppState) => {
    return {
      repo: state.appData.dirName,
      mode: state.appData.mode
    };
  }
)(App);

export default ConnectedApp;
