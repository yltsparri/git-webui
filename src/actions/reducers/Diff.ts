/*
 * Copyright 2017 Ãœlo Parri
 * Copyright 2015 Eric ALBER
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { AnyAction } from "redux";
import Actions from "../Actions";
import { CommitDiff } from "../AppState";
import { CommitViewMode } from "../Commit";
import { FileDiff, Hunk, HunkPart, HunkPartType } from "../git/Diff";

const filter = (hunks: Hunk[]) => {
  const removedHunks = new Array<Hunk>();
  const addedHunks = new Array<Hunk>();
  hunks.forEach(hunk => {
    const leftHunk = {
      header: hunk.header,
      parts: new Array<HunkPart>()
    };
    const rightHunk = {
      header: hunk.header,
      parts: new Array<HunkPart>()
    };
    removedHunks.push(leftHunk);
    addedHunks.push(rightHunk);

    for (let partIndex = 0; partIndex < hunk.parts.length; partIndex++) {
      const part = hunk.parts[partIndex];
      if (part.type === HunkPartType.Keep) {
        leftHunk.parts.push(part);
        rightHunk.parts.push(part);
      } else if (part.type === HunkPartType.Remove) {
        if (partIndex < hunk.parts.length - 1) {
          const next = hunk.parts[partIndex + 1];
          if (next.type === HunkPartType.Add) {
            partIndex++;
            leftHunk.parts.push(part);
            rightHunk.parts.push(next);
            if (next.content.length > part.content.length) {
              leftHunk.parts.push({
                type: null,
                content: new Array<string>(
                  next.content.length - part.content.length + 1
                )
              });
            } else if (next.content.length < part.content.length) {
              rightHunk.parts.push({
                type: null,
                content: new Array<string>(
                  part.content.length - next.content.length + 1
                )
              });
            }
          } else {
            leftHunk.parts.push(part);
            rightHunk.parts.push({
              type: null,
              content: new Array<string>(part.content.length + 1)
            });
          }
        } else {
          rightHunk.parts.push({
            type: null,
            content: new Array<string>(part.content.length + 1)
          });
        }
      } else if (part.type === HunkPartType.Add) {
        leftHunk.parts.push({
          type: null,
          content: new Array<string>(part.content.length + 1)
        });
        rightHunk.parts.push(part);
      }
    }
  });
  return { removedHunks, addedHunks };
};

const getSplitDiff = (fileDiffs: FileDiff[], selectedFile?: number) => {
  if (selectedFile === undefined || !fileDiffs[selectedFile]) {
    return { removedLinesDiff: undefined, addedLinesDiff: undefined };
  }
  const file = fileDiffs[selectedFile];
  const filtered =
    file && file.hunks
      ? filter(file.hunks)
      : { removedHunks: [], addedHunks: [] };
  const removedLinesDiff: FileDiff = {
    hunks: filtered.removedHunks,
    header: file.header,
    fileModeLine: file.fileModeLine,
    indexLine: file.indexLine,
    initialFile: file.initialFile,
    resultingFile: file.resultingFile
  };
  const addedLinesDiff: FileDiff = {
    hunks: filtered.addedHunks,
    header: file.header,
    fileModeLine: file.fileModeLine,
    indexLine: file.indexLine,
    initialFile: file.initialFile,
    resultingFile: file.resultingFile
  };
  return { removedLinesDiff, addedLinesDiff };
};

export function commitDiff(state: CommitDiff, action: AnyAction): CommitDiff {
  switch (action.type) {
    case Actions.UPDATE_COMMIT_DIFF_DATA:
      const splitDiff = getSplitDiff(
        action.data.fileDiffs,
        state.selectedFile || 0
      );
      return {
        ...state,
        ...action.data,
        ...splitDiff,
        selectedFile: action.selectedFile || 0
      };
    case Actions.SELECT_COMMIT:
      return {
        ...state,
        selectedFile: undefined,
        removedLinesDiff: undefined,
        addedLinesDiff: undefined
      };
    case Actions.SELECT_COMMIT_VIEW_MODE:
      return {
        ...state,
        useSplitDiff: action.mode === CommitViewMode.SidebySideDiff
      };
    case Actions.SELECT_COMMIT_DIFF_FILE:
      return {
        ...state,
        selectedFile: action.selectedFile as number,
        ...getSplitDiff(state.fileDiffs, action.selectedFile)
      };
  }
  return (
    state || {
      fileDiffs: new Array<FileDiff>(),
      selectedFile: 0,
      headerLines: new Array<string>(),
      removedLinesDiff: null,
      addedLinesDiff: null,
      useSplitDiff: false
    }
  );
}
