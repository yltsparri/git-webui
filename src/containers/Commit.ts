import { connect } from "react-redux";
import { AnyAction, Dispatch } from "redux";
import DiffActions from "../actions/actioncreators/Diff";
import NavigationActions from "../actions/actioncreators/Navigation";
import TreeActions from "../actions/actioncreators/Tree";
import { AppMode, AppState } from "../actions/AppState";
import FileInfo from "../actions/git/FileInfo";
import CommitView, {
  CommitViewActionProps,
  CommitViewStateProps
} from "../components/CommitView";

const mapStateToProps = (state: AppState): CommitViewStateProps => {
  return {
    diff: state.commitDiff.fileDiffs,
    headerLines: state.commitDiff.headerLines,
    ignoreWhitespace: state.diffOptions.ignoreWhitespace,
    diffContext: state.diffOptions.context,
    fullFileDiff: state.diffOptions.fullFile,
    diffViewMode: state.commits.viewMode,
    path: state.commitTree.path,
    files: state.commitTree.files
  };
};

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    toggleIgnoreWhiteSpace: () => {
      dispatch((DiffActions.toggleIgnoreWhiteSpace() as {}) as AnyAction);
    },
    updateDiffContext: (diffContext: number) => {
      dispatch((DiffActions.setDiffContext(diffContext) as {}) as AnyAction);
    },
    onNodeSelected: (node: FileInfo) => {
      dispatch((TreeActions.selectNode(node) as {}) as AnyAction);
    },
    toggleShowFullFile: () => {
      dispatch((DiffActions.toggleShowFullFile() as {}) as AnyAction);
    },
    onExloreClicked: () => {
      dispatch(NavigationActions.changeAppMode(AppMode.Explore));
    }
  };
};

export default connect<CommitViewStateProps, CommitViewActionProps, any>(
  mapStateToProps,
  mapDispatchToProps
)(CommitView);
