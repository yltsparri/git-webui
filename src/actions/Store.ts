import {applyMiddleware, combineReducers, createStore} from 'redux';
import {appData} from './reducers/AppData';
import {diffOptions} from './reducers/DiffOptions';
import {getBranchDataReducer} from './reducers/Branches';
import {commits} from './reducers/Commits';
import {commitDiff} from './reducers/Diff';
import {commitTree} from './reducers/CommitTree';
import {offsets} from './reducers/Offsets';
import Navigation from './reducers/Navigation';
import Actions from './Actions';
import thunk from 'redux-thunk';

let reducer = combineReducers({
  appData: appData,
  diffOptions: diffOptions,
  localBranches: getBranchDataReducer(Actions.SET_LOCAL_BRANCHES),
  remoteBrances: getBranchDataReducer(Actions.SET_REMOTE_BRANCHES),
  tags: getBranchDataReducer(Actions.SET_TAGS),
  commits,
  offsets,
  commitDiff: commitDiff,
  commitTree: commitTree,
  navigation: Navigation
});

let store = createStore(reducer, applyMiddleware(
  thunk, // lets us dispatch() functions
));

export default store;
