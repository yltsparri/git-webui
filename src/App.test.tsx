import * as React from "react";
import * as ReactDOM from "react-dom";
import { Provider } from "react-redux";
import { create } from "react-test-renderer";
import { AnyAction, applyMiddleware, createStore, Store } from "redux";
import thunk from "redux-thunk";
import {
  AppMode,
  AppState,
  CommitInfo,
  Commits,
  CommitViewMode,
  FileInfo,
  Navigation,
  NavigationNode,
  NavigationType
} from "./actions/AppState";
import { Circle, CommitDiff, DiffOptions, Path } from "./actions/Commit";
import { FileDiff } from "./actions/git/Diff";
import { store } from "./actions/Store";
import { ConnectedApp } from "./App";

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(
    <Provider store={store}>
      <ConnectedApp />
    </Provider>,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});

it("matches snapshot", () => {
  const rootNodes = [
    NavigationType.Workspace,
    NavigationType.RemoteAccess,
    NavigationType.LocalBranches,
    NavigationType.RemoteBranches,
    NavigationType.Tags
  ] as NavigationType[];
  const nodes = {
    [NavigationType.Workspace]: {
      id: NavigationType.Workspace,
      text: "Workspace",
      children: [],
      parentId: "",
      props: {}
    },
    [NavigationType.RemoteAccess]: {
      id: NavigationType.RemoteAccess,
      text: "Remote",
      children: [],
      parentId: "",
      props: {}
    },
    [NavigationType.LocalBranches]: {
      id: NavigationType.LocalBranches,
      text: "Local Branches",
      children: [],
      parentId: "",
      props: {}
    },
    [NavigationType.RemoteBranches]: {
      id: NavigationType.RemoteBranches,
      text: "Remote Branches",
      children: [],
      parentId: "",
      props: {}
    },
    [NavigationType.Tags]: {
      id: NavigationType.Tags,
      text: "Tags",
      children: [],
      parentId: "",
      props: {}
    }
  } as {
    [id: string]: NavigationNode;
  };
  const navigation: Navigation = {
    rootNodes,
    nodes,
    showAll: "",
    selected: ""
  } as Navigation;
  const commits: Commits = {
    commits: new Array<CommitInfo>(),
    selectedCommit: "",
    viewMode: CommitViewMode.Tree,
    graph: { paths: new Array<Path>(), circles: new Array<Circle>() }
  };
  const commitDiff: CommitDiff = {
    fileDiffs: new Array<FileDiff>(),
    selectedFile: 0,
    headerLines: new Array<string>(),
    useSplitDiff: false,
    removedLinesDiff: {} as FileDiff,
    addedLinesDiff: {} as FileDiff
  };
  const state = {
    appData: {
      dirName: "dir",
      viewOnly: true,
      mode: AppMode.History,
      messages: "",
      showAll: "show?"
    },
    navigation,
    commits,
    commitDiff,
    diffOptions: {} as DiffOptions,
    commitTree: {
      path: new Array<FileInfo>(),
      files: new Array<FileInfo>()
    }
  } as AppState;
  const store2 = createStore(
    (s: AppState | undefined) => {
      return s || state;
    },
    state,
    applyMiddleware(
      thunk // lets us dispatch() functions
    )
  ) as Store<AppState, AnyAction>;
  store2.getState();

  const component = create(
    <Provider store={store2}>
      <ConnectedApp />
    </Provider>
  );
  expect(component).toMatchSnapshot();
});
