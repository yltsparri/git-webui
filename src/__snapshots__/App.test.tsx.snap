// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="root"
>
  <div
    id="sidebar"
  >
    <a
      data-target="#help-modal"
      data-toggle="modal"
      href="#"
    >
      <img
        id="sidebar-logo"
        src="/img/git-logo.png"
      />
    </a>
    <div
      id="sidebar-content"
    >
      <section
        id="sidebar-workspace"
      >
        <h4
          onClick={[Function]}
        >
          Workspace
        </h4>
        <ul />
      </section>
      <section
        id="sidebar-remote-access"
      >
        <h4
          onClick={[Function]}
        >
          Remote
        </h4>
        <ul />
      </section>
      <section
        id="sidebar-local-branches"
      >
        <h4
          onClick={[Function]}
        >
          Local Branches
        </h4>
        <ul />
      </section>
      <section
        id="sidebar-remote-branches"
      >
        <h4
          onClick={[Function]}
        >
          Remote Branches
        </h4>
        <ul />
      </section>
      <section
        id="sidebar-tags"
      >
        <h4
          onClick={[Function]}
        >
          Tags
        </h4>
        <ul />
      </section>
    </div>
  </div>
  <div
    id="main-view"
  >
    <div
      id="history-view"
    >
      <div
        id="log-view"
      >
        <div>
          <svg
            height={0}
          />
        </div>
      </div>
      <div
        id="commit-view"
        style={
          Object {
            "display": "flex",
          }
        }
      >
        <div
          id="commit-view-header"
        >
          <ul
            className="nav nav-pills nav-justified"
            role="tablList"
          >
            <li
              className=""
              onClick={[Function]}
            >
              <a
                href="#"
              >
                Commit
              </a>
            </li>
            <li
              className="active"
              onClick={[Function]}
            >
              <a
                href="#"
              >
                Tree
              </a>
            </li>
          </ul>
        </div>
        <div
          className="diff-view-container panel panel-default"
        >
          <div
            className="panel-body"
          >
            <div
              id="tree-view"
            >
              <ol
                className="breadcrumb"
              />
              <div
                className="list-group"
                id="tree-view-tree-content"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
